File: platform/lib/domain/src/main/kotlin/io/hamal/lib/domain/request/Func.kt
package io.hamal.lib.domain.request

import io.hamal.lib.common.value.ValueCode
import io.hamal.lib.domain._enum.CodeType
import io.hamal.lib.domain.vo.*


interface FuncCreateRequest {
    val name: FuncName
    val inputs: FuncInputs
    val code: ValueCode
    val codeType: CodeType
}

data class FuncCreateRequested(
    override val requestId: RequestId,
    override val requestedBy: AuthId,
    override var requestStatus: RequestStatus,
    val id: FuncId,
    val workspaceId: WorkspaceId,
    val namespaceId: NamespaceId,
    val name: FuncName,
    val inputs: FuncInputs,
    val codeId: CodeId,
    val code: ValueCode,
    val codeType: CodeType
) : Requested()

interface FuncUpdateRequest {
    val name: FuncName?
    val inputs: FuncInputs?
    val code: ValueCode?
}

data class FuncUpdateRequested(
    override val requestId: RequestId,
    override val requestedBy: AuthId,
    override var requestStatus: RequestStatus,
    val id: FuncId,
    val workspaceId: WorkspaceId,
    val name: FuncName?,
    val inputs: FuncInputs?,
    val code: ValueCode?
) : Requested()


interface FuncInvokeRequest {
    val correlationId: CorrelationId?
    val inputs: InvocationInputs?
    val version: CodeVersion?
}

interface FuncDeployRequest {
    val version: CodeVersion?
    val message: DeployMessage?
}

data class FuncDeployRequested(
    override val requestId: RequestId,
    override val requestedBy: AuthId,
    override var requestStatus: RequestStatus,
    val id: FuncId,
    val workspaceId: WorkspaceId,
    val version: CodeVersion?,
    val message: DeployMessage?
) : Requested()

File: app/candy-rocks/src/main/frontend/src/types/func.ts
export interface FuncCreateRequested {
    requestId: string;
    requestStatus: string;
    id: string;
    namespaceId: string;
    workspaceId: string;
}

export interface FuncUpdateRequested {
    requestId: string;
    requestStatus: string;
    id: string;
}


export interface FuncInvokeRequested {
    requestId: string;
    requestStatus: string;
    id: string;
}

export interface FuncDeployRequested {
    requestId: string;
    requestStatus: string;
    id: string;
}

export interface Func {
    id: string;
    name: string;
    inputs: object;
    code: {
        id: string;
        version: number;
        value: string;
    },
    deployment: {
        id: string;
        version: number;
        value: string;
        message: string;
    }
}

export interface FuncList {
    funcs: Array<FuncListItem>;
}

export interface FuncListItem {
    id: string;
    name: string;
}

export interface FuncDeploymentList {
    deployments: Array<Deployment>
}

export interface Deployment {
    version: string,
    message: string,
    deployedAt: string
}

